# Most Profitable App Profiles

##  This is a program to establish the most profitable apps as part of our "company"'s published iOS and Android. It's the first project in Dataquest's Data Engineering track

In [1]:
import pandas as pd

In [19]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [20]:
ls

AppleStore.csv                googleplaystore.csv
Profitable App Profile.ipynb


In [50]:
ios_ds = pd.read_csv('AppleStore.csv')
android_ds = pd.read_csv('googleplaystore.csv')

In [88]:
ios_ds.head()

id               track_name  size_bytes currency  price  \
0  284882215                 Facebook   389879808      USD    0.0   
1  389801252                Instagram   113954816      USD    0.0   
2  529479190           Clash of Clans   116476928      USD    0.0   
3  420009108               Temple Run    65921024      USD    0.0   
4  284035177  Pandora - Music & Radio   130242560      USD    0.0   

   rating_count_tot  rating_count_ver  user_rating  user_rating_ver      ver  \
0           2974676               212          3.5              3.5     95.0   
1           2161558              1289          4.5              4.0    10.23   
2           2130805               579          4.5              4.5  9.24.12   
3           1724546              3842          4.5              4.0    1.6.2   
4           1126879              3594          4.0              4.5    8.4.1   

  cont_rating        prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
0          4+  Social Networking               37                1        29   
1         12+      Photo & Video               37                0        29   
2          9+              Games               38                5        18   
3          9+              Games               40                5         1   
4         12+              Music               37                4         1   

   vpp_lic  
0        1  
1        1  
2        1  
3        1  
4        1

In [87]:
android_ds.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

The Kaggle discussion forum has a comment highlighting that row 10472 has some missing data, resulting in data being shifted one column over

In [90]:
android_ds.iloc[[10472]] #Unfortunately, I reran this cell after dropping the row in question below, so this is showing row 10473 instead of 10472

App Category  Rating Reviews  Size     Installs  \
10473  osmino Wi-Fi: free WiFi    TOOLS     4.2  134203  4.1M  10,000,000+   

       Type Price Content Rating Genres    Last Updated Current Ver  \
10473  Free     0       Everyone  Tools  August 7, 2018     6.06.14   

      Android Ver  
10473  4.4 and up

Let's delete that row

In [84]:

android_ds = android_ds.drop(android_ds.index[[10472]])

In [85]:
android_ds.iloc[[10472]]

App Category  Rating Reviews  Size     Installs  \
10473  osmino Wi-Fi: free WiFi    TOOLS     4.2  134203  4.1M  10,000,000+   

       Type Price Content Rating Genres    Last Updated Current Ver  \
10473  Free     0       Everyone  Tools  August 7, 2018     6.06.14   

      Android Ver  
10473  4.4 and up

In [102]:
for idx, row in android_ds.iterrows():
    name = row['App']
    if name == 'Instagram':
        print(row, '\n')

App                        Instagram
Category                      SOCIAL
Rating                           4.5
Reviews                     66577313
Size              Varies with device
Installs              1,000,000,000+
Type                            Free
Price                              0
Content Rating                  Teen
Genres                        Social
Last Updated           July 31, 2018
Current Ver       Varies with device
Android Ver       Varies with device
Name: 2545, dtype: object 

App                        Instagram
Category                      SOCIAL
Rating                           4.5
Reviews                     66577446
Size              Varies with device
Installs              1,000,000,000+
Type                            Free
Price                              0
Content Rating                  Teen
Genres                        Social
Last Updated           July 31, 2018
Current Ver       Varies with device
Android Ver       Varies with device
Name: 2604

Above we see that Instagram appears 4 times in the app. I wonder how many duplicate apps we have in the store. 

In [106]:
duplicate_apps = []
unique_apps = []
for _, app in android_ds.iterrows():
    name = app['App']
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps), '\n')
print('Examples of duplicate apps:', duplicate_apps[:10])

Number of duplicate apps: 1181 

Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


Let's make sure that there really are duplicats of some of the sample apps above. We'll use 'Google My Business', 'ZOOM Cloud Meetings', and 'Slack

In [109]:
gmb = 'Google My Business'
zoom = 'ZOOM Cloud Meetings'
slack = 'Slack'
dupes = [gmb, zoom, slack]

for _, row in android_ds.iterrows():
    if row['App'] in dupes:
        print(row, '\n')
        

App               Google My Business
Category                    BUSINESS
Rating                           4.4
Reviews                        70991
Size              Varies with device
Installs                  5,000,000+
Type                            Free
Price                              0
Content Rating              Everyone
Genres                      Business
Last Updated           July 24, 2018
Current Ver         2.19.0.204537701
Android Ver               4.4 and up
Name: 193, dtype: object 

App               ZOOM Cloud Meetings
Category                     BUSINESS
Rating                            4.4
Reviews                         31614
Size                              37M
Installs                  10,000,000+
Type                             Free
Price                               0
Content Rating               Everyone
Genres                       Business
Last Updated            July 20, 2018
Current Ver            4.1.28165.0716
Android Ver                4.0 and u

Instead of simply removing duplicates at random, we'll use the number of reviews to decide which row for duplicate apps to keep in our data set. This makes intuitive sense since users tend to be drawn toward the apps with the highest number of reviews.

In [113]:
# create a dict to hold our duplicate apps and their number of reviews
reviews_max = {}
for idx, row in android_ds.iterrows():
    name = row['App']
    n_reviews = float(row['Reviews']) 
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    
    if name not in reviews_max:
        reviews_max[name] = n_reviews

In [114]:
len(reviews_max)

9659

Now that we know we have only 9659 unique apps, we're going to create a new data frame, using the dictionary above and our original data set to determine which rows (i.e., the unique apps or the duplicate apps with the most reviews) we want to append to our new data frame.

We'll start by creating a new data frame with our existing column names, and iterating over our old data set, appending only when we find an app that belongs to our reviews_max row

In [115]:
col_names = list(android_ds.columns)

In [117]:
col_names

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [151]:
clean_android_apps = pd.DataFrame(columns=col_names)
already_added = []

In [152]:
print(clean_android_apps)

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []


In [153]:
for _, row in android_ds.iterrows():
    name = row['App']
    n_reviews = float(row['Reviews'])
    if reviews_max[name] == n_reviews and name not in already_added:
        clean_android_apps = clean_android_apps.append(row)
        already_added.append(name)

In [154]:
len(clean_android_apps)

9659

Next, let's define remove any apps not intended for an English-speaking audience. We can use Python's ```ord()``` function to tranpose a character into a given integer, and we know that english-characters are in the range of 0-127. So we'll write a function to determine whether a given app name is english-oriented or not. We'll return ```True``` or ```False``` accordingly and use that to filter out rowns from our dataframe in order to retain an English-oriented dataset.

Additionally, we'll make sure that we're only filtering out app names in our function if there are more than 4 non-english characters. This is because the 0-127 range above only captures the most standard of english characters, including lowercase and uppercase letters, but not emjois or symbols like "™". While we could create a more elaborate function to deal with these use cases, we'll keep it simple for now. The end result should still be fairly effective without requiring a much more involved function.

In [158]:
def is_english_word(app_name):
    non_english_count = 0
    for char in app_name:
        char_int = ord(char)
        if char_int > 127:
            non_english_count += 1
            if non_english_count > 4:
                return False
    return True

In [159]:
print(is_english_word('Instagram'))
print(is_english_word('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english_word('Docs To Go™ Free Office Suite'))
print(is_english_word('Instachat 😜'))

True
False
True
True


In [163]:
english_android_apps = pd.DataFrame(columns=col_names)
print(clean_english_ds)

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []


In [164]:
for _, row in clean_android_apps.iterrows():
    name = row['App']
    if is_english_word(name):
        english_android_apps = english_android_apps.append(row)

In [165]:
len(english_android_apps)

9619

Our last step is to isolate the free apps. We're only looking at revenue in apps that will have in-app purchases (i.e., non-paid apps) so we want to work with only those apps.

In [169]:
english_android_apps.sample(50)

App             Category  \
9424   Esporte Interativo - Notícias e Resultados Ao ...               SPORTS   
1878                                        Clash Royale                 GAME   
4938                                           Pocket AC          PHOTOGRAPHY   
7437                             China Town Auto Theft 2                 GAME   
5771                                             News.aw   NEWS_AND_MAGAZINES   
2482                               Familyfirst Messenger              MEDICAL   
6158                       BGKontakti Bayern BG Kontakti               SOCIAL   
1941                                        Major Mayhem                 GAME   
4693                                       V.360° Camera          PHOTOGRAPHY   
10750                            Finger Scanner Gestures                TOOLS   
7358                                        swissinfo.ch   NEWS_AND_MAGAZINES   
3289                               Tiny Flashlight + LED                TOOLS   
7081                                    iBiznes24 mobile              FINANCE   
10058                            Moonlight GO Weather EX              WEATHER   
7744                               Sniper Killer Shooter                 GAME   
7694                                 CP Smart Check List      PERSONALIZATION   
1221                      ChefTap Recipes & Grocery List       FOOD_AND_DRINK   
3931                                         Like A Boss               FAMILY   
6291                                             MinT BI         PRODUCTIVITY   
4913                          Trane Interactive Smart AC                TOOLS   
8159                                           CZ Kompas                TOOLS   
4548                          R Quick Reference Big Data  BOOKS_AND_REFERENCE   
3294                                   Flashlight HD LED                TOOLS   
2339                                           MoodSpace              MEDICAL   
3961                                              B.GOOD       FOOD_AND_DRINK   
1127                                         Stock Quote              FINANCE   
8992                                            DW Audio               FAMILY   
5589                              Ninja Turtles: Legends               FAMILY   
1815                         TerraGenesis - Space Colony                 GAME   
8324                                    DF Night Selfies          PHOTOGRAPHY   
2296                       EMT Tutor NREMT-B Study Guide              MEDICAL   
2384                       Deep Sleep and Relax Hypnosis              MEDICAL   
4162                                            G-Switch                 GAME   
3189                        Cheapflights – Flight Search     TRAVEL_AND_LOCAL   
6715   True Skateboarding Ride Skateboard Game Freestyle                 GAME   
7638                              Aquarium Co-Op Podcast            LIFESTYLE   
5048                                 AppFinder by AppTap                TOOLS   
5859                          Display Phone Screen On TV                TOOLS   
2108                          Paw Patrol by ShuffleCards               FAMILY   
3875                                        A Word A Day               FAMILY   
7567                      Vintage Flower CM Locker Theme      PERSONALIZATION   
6612                      BP Tracker-Symptoms & Solution              MEDICAL   
1104                              Chime - Mobile Banking              FINANCE   
9608                                 Yoong: Kick 'Em Up!               SPORTS   
5601                                          Google Pay              FINANCE   
597                        Luxy Pro- Elite Dating Single               DATING   
10235     Unlimited Group Links - Whatsapp, FB, Telegram               SOCIAL   
8437                      Shred! Downhill Mountainbiking                 GAME   
6147                                      BG TV Stations               FAMILY   
608

In [170]:
free_english_android_apps = pd.DataFrame(columns=col_names)
print(free_english_android_apps)

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []


In [171]:
for _, row in english_android_apps.iterrows():
    if row['Type'] == 'Free':
        free_english_android_apps = free_english_android_apps.append(row)

In [172]:
len(free_english_android_apps)

8868

Next, we'll use the 'Genres' and 'Category' columns in the android dataset and the 'prime_genre' category in the ios dataset to determine which types of apps tend to be the most popular. We'll do this by creating a frequency table for each of those categories.

We'll define a function to do this for each of the genres we're interested in. The function will take in a dataset, and the column name we're interested in. It'll create a dictionary with the the discrete column values as the key and the count of apps under that column. Since dictionaries aren't all that useful for sorting, we'll create a list of tuples (each tuple being a key-value pair from that dictionary) and sort the list. This sorted tuple list is the frequency table that our function will return.